### PR TITLE
Swagger addon

### PR DESCRIPTION
**Pull request by maria-robobug**
This PR adds Swagger documentation to the backend API by using the [flasgger](https://pypi.org/project/flasgger/) library.
It allows us to automatically generate swagger documentation and run the swagger-ui at `localhost:8080/apidocs`- the Swagger json is also available from that page.

I've also updated the `docker-compose.yml` to use wait-for-it.sh and `depends_on` as discussed.

### CODE REVIEW

File: travel.py [hunk 1]
```diff
@@ -66,13 +88,41 @@ def make_user_key(username):
...
     """Airport class for airport objects in the database"""
 
     @api.route('/airports', methods=['GET', 'OPTIONS'])
     @cross_origin(supports_credentials=True)
     def airports():
-        """Returns list of matching airports and the source query"""
+        """Returns list of matching airports and the source query
+        ---
+        tags:
+        - aiports
```
Comment (osfameron) [inline travel.py]: typo

Comment (maria-robobug) [reply travel.py]: aiports lol, 👍

File: travel.py [hunk 2]
```diff
@@ -97,15 +147,65 @@ def airports():
...
+                        example: [{
+                                    "destinationairport": "LAX",
+                                    "equipment": "738",
+                                    "flight": "AA331",
+                                    "flighttime": 1220,
+                                    "name": "American Airlines",
+                                    "price": 152.5,
+                                    "sourceairport": "SFO",
+                                    "utc": "16:37:00"
+                                }]
```
Comment (osfameron) [inline travel.py]: indentation

File: travel.py [hunk 3]
```diff
@@ -150,13 +250,68 @@ def flightPaths(fromloc, toloc):
...
+              description: Returns login data and query context information
+              content:
+                application/json:
+                  schema:
+                    type: object
+                    properties:
+                      context:
+                        type: array
+                        format: string
+                        example: ["A description of a KV operation"]
```
Comment (osfameron) [inline travel.py]: Would be bit more work, but might be nice for the example Context outputs to give the most common (happy path) output?

File: travel.py [hunk 4]
```diff
@@ -182,14 +337,69 @@ def login(tenant):
...
+              description: Returns a conflict error
+              content:
+                application/json:
+                    schema:
+                       type: object
+                       properties:
+                        message:
+                           type: string
+                           example: ["User already exists"]
+        """
```
Comment (osfameron) [inline travel.py:418]: OK, so seeing these ~50 lines repeated from above does actually motivate figuring out the `$ref` stuff doesn't it? 😂

File: travel.py [hunk 5]
```diff
@@ -214,80 +424,253 @@ def signup(tenant):
...
+                        type: array
+                        format: object
+                        example: [{
+                                    "date": "12/12/2020",
+                                    "destinationairport": "Leonardo Da Vinci International Airport",
+                                    "flight": "12RF",
+                                    "name": "boeing",
+                                    "price": 50.0,
+                                    "sourceairport": "London (Gatwick)"
+                                 }]
```
Comment (osfameron) [inline travel.py]: This is great work - I love seeing an example of the output specified.

File: travel.py (hunk 4 above)
Comment (maria-robobug) [reply travel.py:418]: indeed, so hard to keep up with all of this comment, indentation hell 😆

File: travel.py [hunk 6]
```diff
@@ -214,80 +424,253 @@ def signup(tenant):
...
+                       type: object
+                       properties:
+                        message:
+                           type: string
+                           example: ["User does not exist"]
+        security:
+            - bearer: []
+        """
+        agent = make_user_key(tenant)
+        user = make_user_key(username)
```
Comment (osfameron) [inline travel.py:601]: thanks Git, for assuming that these blank lines are useful context and completely misinterpreting the diff... *sigh*

File: travel.py [hunk 7]
```diff
@@ -84,8 +84,8 @@ def index():
     return jsonify(endpoints)
 
 
-def make_user_key(username):
-    return username.lower()
+def lowercase(key):
```
Comment (osfameron) [inline travel.py:136]: NB: No calling code modified in this commit.

File: travel.py [hunk 8]
```diff
@@ -312,9 +305,9 @@ def login(tenant):
                            type: string
                            example: ["Password does not match"]
         """
-        agent = make_user_key(tenant)
+        agent = lowercase(tenant)
```
Comment (osfameron) [inline travel.py:342]: aha, caller modified in subsequent commit, ok.

File: travel.py [hunk 9]
```diff
@@ -312,9 +305,9 @@ def login(tenant):
...
         req = request.get_json()
-        user = make_user_key(req['user'])
+        user = lowercase(req['user'])
```
Comment (osfameron) [inline travel.py:344]: not a biggie, but at this point why not just use `.lower()` directly?
I think in the JS code I kept `makeKey` as it implies that it's a method for normalizing keys (even if the implementation itself is then a trivial placeholder...)

File: travel.py [hunk 10]
```diff
@@ -77,13 +77,6 @@
 CORS(app, headers=['Content-Type', 'Authorization'])
 
 
-@app.route('/')
```
Comment (osfameron) [inline travel.py:58]: Adieu faithful endpoint list! 😭

File: travel.py [hunk 11]
```diff
@@ -298,12 +309,8 @@ def login(tenant):
...
               content:
                 application/json:
-                    schema:
-                       type: object
-                       properties:
-                        message:
-                           type: string
-                           example: ["Password does not match"]
+                    schema: 
+                      $ref: '#/components/schemas/ErrorModel'
```
Comment (osfameron) [inline travel.py]: nice!

**Comment by osfameron** [comment]
Looks good! 👍

File: travel.py (hunk 3 above)
Comment (maria-robobug) [reply travel.py]: Yeah can do, especially when I extract this into a common object for reuse. We will have to pick a common example just to show something.

Obviously with the UI you can actually test the endpoint so you will see what the real response looks like anyways.

File: travel.py (hunk 10 above)
Comment (maria-robobug) [reply travel.py:58]: 👋

File: travel.py (hunk 9 above)
Comment (maria-robobug) [reply travel.py:344]: yeah I kept it but not entirely convinced it's totally necessary, .lower() on its own would have been enough, but I think it should be fine to leave for now.

File: travel.py [hunk 12]
```diff
@@ -77,6 +77,29 @@
                         "example": ["An error message"]
                     }
                 }
+            },
+            "ContextModel": {
+                "type": "array",
+                "items": {"type": "string"}
+            },
+            "ResultDataListModel": {
```
Comment (maria-robobug) [inline travel.py]: I am not too sure about the naming here, 🆘 !

File: travel.py [hunk 13]
```diff
@@ -77,6 +77,29 @@
...
+                "type": "object",
+                "properties": {
+                    "context": {"$ref": "#/components/schemas/ContextModel"},
+                    "data": {
+                        "type": "array",
+                        "items": {"type": "object"}
+                    }
+                }
+            },
+            "ResultDataObjectModel": {
```
Comment (maria-robobug) [inline travel.py]: Same here 😭

File: travel.py (hunk 12 above)
Comment (osfameron) [reply travel.py]: I think the name is fine.
But perhaps `ResultList` vs `ResultSingleton` ?

File: travel.py [hunk 14]
```diff
@@ -52,7 +52,7 @@
 app.config.from_object(__name__)
 app.config['CORS_HEADERS'] = 'Content-Type'
 app.config['SWAGGER'] = {
-    'openapi': '3.0.5',
+    'openapi': '3.0.3',
```
Comment (maria-robobug) [inline travel.py:54]: Dunno where I got `3.0.5` from 😆  - shame on this library for not telling me about it though.

File: travel.py [hunk 15]
```diff
@@ -631,7 +631,7 @@ def updateflights(tenant, username):
 class HotelView(SwaggerView):
     """Class for storing Hotel search related information"""
 
-    @api.route('/hotels/<description>/<location>', methods=['GET'])
+    @api.route('/hotels/<description>/<location>/', methods=['GET'])
```
Comment (maria-robobug) [inline travel.py:633]: Seems like we want the `/` at the end when calling from the FE, so I'm putting this back - I wasn't sure why we needed it, but in the spirit of not breaking things with FE I will leave it for now.

**Pull request merged**